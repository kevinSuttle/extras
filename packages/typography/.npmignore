docs
test
babel.config.js
public
.cache
