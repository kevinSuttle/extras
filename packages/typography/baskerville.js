module.exports = require('./dist/themes/baskerville')
