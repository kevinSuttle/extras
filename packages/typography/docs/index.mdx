import Root from './root'
import Readme from '../README.md'

export default Root

<Readme />

---

# Typography

## Demo

This section is only meant as a typographic demo and does not contain any API documentation or usage guidelines.

> “This is honestly my favourite way to build UI components right now”

# Hamburger
## Hamburger
### Hamburger
#### Hamburger
##### Hamburger
###### Hamburger

