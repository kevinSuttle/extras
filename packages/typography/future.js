module.exports = require('./dist/themes/future')
