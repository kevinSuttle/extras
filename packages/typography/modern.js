module.exports = require('./dist/themes/modern')
