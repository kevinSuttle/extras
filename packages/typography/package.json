{
  "name": "@styled-system/typography",
  "version": "1.0.0-8",
  "description": "Add typographic styles to child elements using styled-system and themes",
  "main": "dist/index.js",
  "scripts": {
    "prepare": "babel src -d dist",
    "start": "gx docs",
    "test": "jest"
  },
  "keywords": [],
  "author": "Brent Jackson <jxnblk@gmail.com>",
  "license": "MIT",
  "devDependencies": {
    "@babel/cli": "^7.2.3",
    "@babel/core": "^7.3.4",
    "@babel/preset-env": "^7.3.4",
    "@babel/preset-react": "^7.0.0",
    "@emotion/core": "^10.0.9",
    "@emotion/styled": "^10.0.9",
    "emotion-theming": "^10.0.9",
    "gx": "^2.0.0-16",
    "jest": "^24.4.0",
    "react": "^16.8.4",
    "react-dom": "^16.8.4",
    "react-helmet": "^5.2.0",
    "styled-components": "^4.2.0",
    "typography-theme-lincoln": "^0.16.19"
  },
  "dependencies": {
    "lodash.merge": "^4.6.1",
    "lodash.omit": "^4.5.0",
    "lodash.pick": "^4.4.0",
    "styled-system": "^4.0.3",
    "typography": "^0.16.19"
  },
  "jest": {
    "testMatch": [
      "**/test/**/*.js"
    ]
  }
}
