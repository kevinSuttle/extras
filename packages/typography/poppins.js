module.exports = require('./dist/themes/poppins')
