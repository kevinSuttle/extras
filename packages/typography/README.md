
# Styled System Typography

Add typographic styles to child elements using styled-system theme primitives â€“ intended for MDX or markdown content
https://styled-system.com/typography

**EXPERIMENTAL/WIP**

```sh
npm i @styled-system/typography
```

```jsx
import React from 'react'
import { Global } from '@emotion/core'
import { ThemeProvider } from 'emotion-theming'
import typography from '@styled-system/typography'
import theme from './theme'

export default props =>
  <ThemeProvider theme={theme}>
    <>
      <Global styles={typography} />
      {props.children}
    </>
  </ThemeProvider>
```

```js
// example theme.js
export default {
  space: [
    0, 4, 8, 16, 32, 64, 128, 256
  ],
  fontSizes: [
    12, 14, 16, 20, 24, 32, 48, 64, 96, 128
  ],
  // the typography object defines global typography styles
  // based on the base theme scales
  typography {
    body: {
      fontFamily: 'system-ui, sans-serif',
      lineHeight: 1.5,
    },
    h1: {
      fontSize: [5, 6],
      lineHeight: 1.25,
      mt: [3, 4],
      mb: 3,
    },
    h2: {
      fontSize: [4, 5],
      lineHeight: 1.25,
    },
    h3: {
      fontSize: 3,
      lineHeight: 1.25,
    },
    pre: {
      p: [2, 3],
    },
  }
}
```

## Getting Started

Styled System Typography is intended for use with [Emotion][] or [Styled Components][], be sure your build setup includes support
for one of these libraries.
To configure the theme, add a `ThemeProvider` to your application's root.

```jsx
import React from 'react'
import { ThemeProvider } from 'emotion-theming'
import theme from './theme'

export default props =>
  <ThemeProvider theme={theme}>
    {props.children}
  </ThemeProvider>
```

## Theming

The `typography` object in the theme defines the global styles applied by this module. Each key in the object corresponds to an HTML element selector, where many common typographic elements can also use styled-system typographic props to define responsive, theme-based styles.

### Props

[Styled System][] props can pick up values from a theme context when using a `ThemeProvider` component.
This means that the value passed to the prop can be a key from the correlating theme object.
For example, setting `fontSize: 5` picks up the `theme.fontSizes[5]` value, which defaults to `32px`.
All [styled-system][] props also accept arrays as values to create [responsive props][].
Read more about how Styles System works in the [docs][styled-system].

The core typographic elements accept styles using the following [styled-system][] typography props in addition to all CSS properties:

- `fontFamily`
- `fontSize`
- `fontWeight`
- `lineHeight`
- space props:
  - `m`: margin
  - `mt`: margin-top
  - `mr`: margin-right
  - `mb`: margin-bottom
  - `ml`: margin-left
  - `mx`: (x-axis) margin-left and margin-right
  - `my`: (y-axis) margin-top and margin-bottom
  - `p`: padding
  - `pt`: padding-top
  - `pr`: padding-right
  - `pb`: padding-bottom
  - `pl`: padding-left
  - `px`: (x-axis) padding-left and padding-right
  - `py`: (y-axis) padding-top and padding-bottom
- `color`
- `bg` (or `backgroundColor`)

The following elements use thes typographic style props:
`body`, `h1`, `h2`, `h3`, `h4`, `h5`, `h6`, `a`, `p`, `ol`, `ul`, `dl`, `dd`, `li`, `blockquote`, `hr`, `img`, `pre`, `code`, `samp`, `kbd`, `table`, `tr`, `th`, `td`, `b`, `strong`, `em`, `i`, `abbr`


## Built-In Themes

This library includes several themes that can be used out-of-the-box or as the basis for a custom typography theme.

```jsx
// example theme
import { themes } from '@styled-system/typography'

export default {
  typography: themes.future,
}
```

## Scoped Styles

Styled System Typography can be "scoped" to a wrapping `<div>` or other element instead of adding global styles.
This approach **will increase CSS specificity** due to the class selector required, which will make overriding styles within the component more difficult. **Use with caution**.

```jsx
import styled from '@emotion/styled'
import typography from '@styled-system/typography'

const Typography = styled.div(typography)
```

---

### Related

The [typography-system][] module attempts to solve a similar problem with a different implementation. It takes a Typography.js theme and converts it into a theme object that works with Styled System. The component in this package starts with primitive values from a Styled System theme, and creates typographic styles based on those primitives.

### Prior Art

This component is inspired by many other similar projects, especially [Typography.js][]

[styled system]: https://styled-system.com
[styled-system]: https://styled-system.com
[responsive props]: https://styled-system.com/responsive-styles
[emotion]: https://emotion.sh
[styled components]: https://styled-components.com
[typography.js]: https://kyleamathews.github.io/typography.js/
[typography-system]: https://github.com/jxnblk/typography-system

MIT License
