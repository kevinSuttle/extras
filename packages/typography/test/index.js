import typography, { themes } from '../src'

test('returns a styles object', () => {
  const styles = typography({
    theme: {
      typography: themes.future
    }
  })
  expect(styles).toMatchSnapshot()
})
