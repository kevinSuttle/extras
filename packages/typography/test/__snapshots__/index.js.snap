// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns a styles object 1`] = `
Object {
  "&": Object {
    "fontFamily": "\\"Avenir Next\\", system-ui, sans-serif",
    "lineHeight": 1.5,
  },
  "body": Object {
    "fontFamily": "\\"Avenir Next\\", system-ui, sans-serif",
    "lineHeight": 1.5,
  },
  "code": Object {
    "fontFamily": "monospace",
    "fontSize": "85%",
  },
  "h1": Object {
    "@media screen and (min-width: 40em)": Object {
      "fontSize": "48px",
    },
    "fontSize": "32px",
    "fontWeight": 600,
    "letterSpacing": "0.2em",
    "marginBottom": "64px",
    "marginTop": "64px",
    "textTransform": "uppercase",
  },
  "p": Object {
    "@media screen and (min-width: 40em)": Object {
      "fontSize": "20px",
    },
    "fontSize": "16px",
    "lineHeight": 1.625,
    "marginBottom": "32px",
    "marginTop": "32px",
    "maxWidth": "40em",
  },
  "pre": Object {
    "borderRadius": 8,
    "fontFamily": "monospace",
    "overflowX": "auto",
    "padding": "16px",
  },
}
`;
